tutorial/20_recipes/006_user_defined_pruner.py

```diff
@@ -0,0 +1,130 @@
+returns a boolean value: `True` if the trial should be pruned and `False`
+otherwise. Using the Study and Trial objects, you can access all other trials
+through the :meth:`study.get_trial` method and, and from a trial, its reported
+intermediate values through the `trial.intermediate_values` attribute (a
+dictionary which maps an integer `step` to a float value).
+
+You can refer to the source code of the built-in Optuna pruners as templates for
+building your own. In this document, for illustration, we describe the
+construction and usage of a simple (but aggressive) pruner which prunes trials
+that are in last place compared to completed trials at the same step.
+
+.. note::
+  Please refer to the documentation of :class:`~optuna.pruners.BasePruner` or,
+  for example, :class:`~optuna.pruners.ThresholdPruner` or
+  :class:`optuna.pruners.PercentilePruner` for more robust examples of pruner
+  implementation, including error checking and complex pruner-internal logic.
+
+An Example: Implementing ``LastPlacePruner``
+--------------------------------------------
+
+We aim to optimize the ``loss`` and ``alpha`` hyperparameters for a stochastic
+gradient descent classifier (``SGDClassifier``) run on the sklearn iris dataset. We
+implement a pruner which terminates a trial at a certain step if it is in last
+place compared to completed trials at the same step. We begin considering
+pruning after a "warmup" of 1 training step and 5 completed trials. For
+demonstration purposes, we print() a diagnostic message from prune() when it is
+about to return True (indicating pruning).
+
+It may be important to note that the SGDClassifier score, as it is evaluated on
+a holdout set, decreases with enough training steps due to overfitting. This
+means that a trial could be pruned even if it had a favorable (high) value on a
+previous training set. After pruning, Optuna will take the intermediate value
+last reported as the value of the trial.
+
+"""
+
+import numpy as np
+from sklearn.datasets import load_iris
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import SGDClassifier
+
+import optuna
+from optuna.pruners import BasePruner
+from optuna.trial._state import TrialState
+
+
+class LastPlacePruner(BasePruner):
+    def __init__(self, warmup_steps, warmup_trials):
+        self._warmup_steps = warmup_steps
+        self._warmup_trials = warmup_trials
+
+    def prune(self, study: "optuna.study.Study", trial: "optuna.trial.FrozenTrial") -> bool:
+        # Get the latest score reported from this trial
+        step = trial.last_step
+
+        if step:  # trial.last_step == None when no scores have been reported yet
+            this_score = trial.intermediate_values[step]
+
+            # Get scores from other trials in the study reported at the same step
+            all_trials = study.get_trials(deepcopy=False)
+            completed_trials = [t for t in all_trials if t.state == TrialState.COMPLETE]
```

Comment:
We recently introduced a second argument to `Study.get_trials` to filter on states. You can save some compute time and reduce code. It's definitely optional since we still have code all over Optuna like the one you provided.
```suggestion
            completed_trials = study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,))
```
_Suggestion is not verified._

Reply:
@hvy Hello, sorry I haven't worked on this PR in a while. But I noticed that the CI checks fail to run the .py program generated from this PR, since the version of Optuna that they use does not use the updated get_trials() signature with the states= argument.

Reply:
Thanks, noticed that you managed to fix this by merging the master. (c.f. https://circleci.com/api/v1.1/project/github/optuna/optuna/65256/output/104/0?file=true&allocation-id=5ffe921a1611687798db53ce-0-build%2F708C4094)